### PR TITLE
User model and user database relations

### PR DESCRIPTION
**Pull request by mdros**

### CODE REVIEW

File: src/main/java/pl/uj/io/allercheck/model/User.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            joinColumns = { @JoinColumn(name = "user_id") },
+            inverseJoinColumns = { @JoinColumn(name = "product_id") }
+    )
+    private Set<Product> favoriteProducts = new HashSet<>();
+
+    @ManyToMany(cascade = {CascadeType.MERGE})
+    @JoinTable(
+            name = "user_product",
+            joinColumns = { @JoinColumn(name = "user_id") },
+            inverseJoinColumns = { @JoinColumn(name = "product_id") }
```
Comment (kploskonka) [inline src/main/java/pl/uj/io/allercheck/model/User.java:43]: favorite products and products blacklist should not be in the same table "user_product" as there's no way to distinguish them

File: src/main/java/pl/uj/io/allercheck/model/Product.java [hunk 2]
```diff
@@ -44,6 +44,12 @@ public Product(int id, String name) {
...
 
+    @ManyToMany(mappedBy = "favoriteProducts")
+    private Set<User> usersFavorites;
+
+    @ManyToMany(mappedBy = "productBlacklist")
+    private Set<User> usersBlacklists;
+
     public void removeAllergen(Allergen allergen) {
         allergenList.remove(allergen);
     }
```
Comment (kploskonka) [inline src/main/java/pl/uj/io/allercheck/model/Product.java:55]: Add removal strategy, what if an user or a product gets removed? Should be the same case as in Allergens but please check if works.

File: src/main/java/pl/uj/io/allercheck/model/User.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    )
+    private Set<Product> productBlacklist = new HashSet<>();
+
+    public User(String email, String password) {
+        this.email = requireNonNull(email);
+        this.password = requireNonNull(password);
```
Comment (kploskonka) [inline src/main/java/pl/uj/io/allercheck/model/User.java:49]: You can write Objects.requireNonNull() and therefore avoid static import from line 11

**Review by kploskonka — CHANGES_REQUESTED** [review]

File: src/main/java/pl/uj/io/allercheck/model/User.java (hunk 1 above)
Comment (mdros) [reply src/main/java/pl/uj/io/allercheck/model/User.java:43]: Sure, thanks a lot for this insightful message @kploskonka, I'll fix this ASAP

File: src/main/java/pl/uj/io/allercheck/model/User.java (hunk 3 above)
Comment (mdros) [reply src/main/java/pl/uj/io/allercheck/model/User.java:49]: I think that static import looks better, I'd opt for this solution

File: src/main/java/pl/uj/io/allercheck/model/Product.java (hunk 2 above)
Comment (mdros) [reply src/main/java/pl/uj/io/allercheck/model/Product.java:55]: Done!

File: src/main/java/pl/uj/io/allercheck/model/User.java (hunk 3 above)
Comment (kploskonka) [reply src/main/java/pl/uj/io/allercheck/model/User.java:49]: https://stackoverflow.com/questions/420791/what-is-a-good-use-case-for-static-import-of-methods

Perhaps this answer changes your mind

Comment (mdros) [reply src/main/java/pl/uj/io/allercheck/model/User.java:49]: After reading the answers, I think that this case is not a bad example of static import:
I am importing specific method, as the answer suggests to do, and I think that one static import isn't polluting the namespace as we're not importing anything besides that

**Review by kploskonka — APPROVED** [review]

**Pull request merged**